pkg/chann/chann.go

```diff
@@ -213,11 +158,89 @@ func (ch *Chann[T]) Close() {
 	switch ch.cfg.typ {
 	case buffered, unbuffered:
 		close(ch.in)
+		close(ch.close)
 	default:
 		ch.close <- struct{}{}
 	}
 }
 
+// unboundedProcessing is a processing loop that implements unbounded
+// channel semantics.
+func (ch *Chann[T]) unboundedProcessing() {
+	var nilT T
+
+	ch.q = make([]T, 0, 1<<10)
+	for {
+		select {
+		case e, ok := <-ch.in:
+			if !ok {
+				panic("chann: send-only channel ch.In() closed unexpectedly")
+			}
+			atomic.AddInt64(&ch.cfg.len, 1)
+			ch.q = append(ch.q, e)
+		case <-ch.close:
+			ch.unboundedTerminate()
+			return
+		}
+
+		for len(ch.q) > 0 {
+			select {
+			case ch.out <- ch.q[0]:
+				atomic.AddInt64(&ch.cfg.len, -1)
+				ch.q[0] = nilT
+				ch.q = ch.q[1:]
+			case e, ok := <-ch.in:
+				if !ok {
+					panic("chann: send-only channel ch.In() closed unexpectedly")
+				}
+				atomic.AddInt64(&ch.cfg.len, 1)
+				ch.q = append(ch.q, e)
+			case <-ch.close:
+				ch.unboundedTerminate()
+				return
+			}
+		}
+		if cap(ch.q) < 1<<5 {
+			ch.q = make([]T, 0, 1<<10)
+		}
+	}
+}
+
+// unboundedTerminate terminates the unbounde channel's processing loop
+// and make sure all unprocessed elements either be consumed if there is
+// a pending receiver.
+func (ch *Chann[T]) unboundedTerminate() {
+	var nilT T
+
+	close(ch.in)
+	for e := range ch.in {
+		ch.q = append(ch.q, e)
+	}
+	for len(ch.q) > 0 {
+		select {
+		case ch.out <- ch.q[0]:
+		// The default branch exists because we need guarantee
+		// the loop can terminate. If there is a receiver, the
+		// first case will ways be selected. See https://github.com/golang-design/chann/issues/3.
+		default:
```

Comment:
Is it possible to cause data loss here?  Or we don't care if the data is consumed properly after chan is closed?

Reply:
We don't care about data loss, it's more about deleting that data and gc as soon as possible here.

Reply:
After chan closed, slow consumers can actually only consume part of their data, which is pretty confusing.
So why not just close `out`, which provides a more explicit guarantee of the semantics.

Reply:
Would this be less friendly to consumers who can handle it quickly?

Reply:
No matter how fast consumers consume, there is always the risk of losing data, which can be quite insidious and dangerous. In this sense, it is always safer to lose a continuous piece of data than to lose a random piece of data.

Reply:
Removed the default. 
Thanks for your review!  💚 💙 💜 💛 ❤️